### PR TITLE
ELEMENTS-1194: release when merging on reference branches

### PR DESCRIPTION
**Pull request by Gabez0r**

### CODE REVIEW

**Comment by Gabez0r** [comment]
Requires nuxeo/jx-webui-env#18.

File: Jenkinsfile [hunk 1]
```diff
@@ -22,23 +22,75 @@ properties([
...
+      statusResultSource: [$class: 'ConditionalStatusResultSource', results: [[$class: 'AnyBuildResult', message: message, state: state]]],
+    ])
+  }
+}
+
+def isPullRequest() {
+  return BRANCH_NAME =~ /PR-.*/
 }
 
+String getPackageVersion() {
```
Comment (nelsonsilva) [inline Jenkinsfile:39]: nit: for consistency we should either use `def` or declare the return type.

File: Jenkinsfile [hunk 2]
```diff
@@ -22,23 +22,75 @@ properties([
...
+  container('nodejs') {
+    return sh(script: 'npx -c \'echo "$npm_package_version"\'', returnStdout: true).trim()
+  }
+}
+
+String getReleaseVersion(String preid) {
+  String currentSnapshot = getPackageVersion()
+  String nextPromotion = currentSnapshot.replace("-SNAPSHOT", "")
```
Comment (nelsonsilva) [inline Jenkinsfile:47]:
```suggestion
  String nextPromotion = getPackageVersion().replace("-SNAPSHOT", "")
```

File: Jenkinsfile [hunk 3]
```diff
@@ -22,23 +22,75 @@ properties([
...
+  String version = "${nextPromotion}-${preid}.0" // first version ever
```
Comment (nelsonsilva) [inline Jenkinsfile:48]: nit: same, either `def` or `String`

File: Jenkinsfile [hunk 4]
```diff
@@ -22,23 +22,75 @@ properties([
...
+  }
+  return version
+}
+
+String getPullRequestVersion() {
+  return getPackageVersion() + "-${BRANCH_NAME}"
+}
+
+String getVersion() {
+  return isPullRequest() ? getPullRequestVersion() : getReleaseVersion('rc')
```
Comment (nelsonsilva) [inline Jenkinsfile:66]: there's really no point in making the `preid` an argument, might as well have it set as a constant and used in `getReleaseVersion()`

File: Jenkinsfile [hunk 5]
```diff
@@ -92,63 +144,117 @@ pipeline {
     }
     stage('Build and deploy preview') {
       when {
-        branch 'PR-*'
+        allOf {
+          branch 'PR-*'
+          not {
+            environment name: 'DRY_RUN', value: 'true'
```
Comment (nelsonsilva) [inline Jenkinsfile:150]: nit: would be nice to move this to a `isDryRun()` if possible (not sure it'll work in these declarative blocks

File: Jenkinsfile [hunk 6]
```diff
@@ -92,63 +144,117 @@ pipeline {
...
+      }
+      steps {
+        container('nodejs') {
+          echo """
+          --------
+          Git commit, tag and push
+          --------
+          """
+          sh """
+            #!/usr/bin/env bash -xe
```
Comment (nelsonsilva) [inline Jenkinsfile:191]: do we need to use bash and set `-xe`? is this not the default ?

File: Jenkinsfile (hunk 5 above)
Comment (Gabez0r) [reply Jenkinsfile:150]: Don't think we can use it here...

File: Jenkinsfile (hunk 6 above)
Comment (Gabez0r) [reply Jenkinsfile:191]: Not sure. Took it from here https://github.com/nuxeo/nuxeo/blob/master/ci/Jenkinsfiles/build.groovy#L644.

File: Jenkinsfile [hunk 7]
```diff
@@ -22,23 +22,75 @@ properties([
...
 }
 
+def getPackageVersion() {
+  container('nodejs') {
+    return sh(script: 'npx -c \'echo "$npm_package_version"\'', returnStdout: true).trim()
+  }
+}
+
+def getReleaseVersion() {
+  def preid = 'rc'
```
Comment (Gabez0r) [inline Jenkinsfile:46]: What pre-release ID should we be using? 'rc' is pretty common, but 'alpha' and 'beta' are also used, and I've seen 'pre' as well. Polymer uses 'rc' for their library and 'pre' for individual elements.

Comment (nelsonsilva) [reply Jenkinsfile:46]: I'd say we stick to "rc" for now since that's what these versions are anyway "release candidates"

**Pull request merged**